File: src/main/java/com/github/hirsivaja/ip/IpProtocol.java
package com.github.hirsivaja.ip;

public sealed interface IpProtocol permits IpProtocol.GenericIpProtocol, IpProtocols {
    byte type();

    static IpProtocol fromType(byte type) {
        for (IpProtocol identifier : IpProtocols.values()) {
            if (identifier.type() == type) {
                return identifier;
            }
        }
        return new GenericIpProtocol(type);
    }

    record GenericIpProtocol(byte type) implements IpProtocol {}
}

File: src/main/kotlin/com/github/hirsivaja/ip/protocol/IpProtocol.kt
package com.github.hirsivaja.ip.protocol

sealed interface IpProtocol {
    val type: UByte

    companion object {
        fun fromType(type: UByte): IpProtocol {
            for (identifier in IpProtocols.entries) {
                if (identifier.type == type) {
                    return identifier
                }
            }
            return GenericIpProtocol(type)
        }
    }

    data class GenericIpProtocol(override val type: UByte) : IpProtocol
}
